File: zatca/zatca/hashing.py
"""Invoice hashing and the per-device hash chain.

Every ZATCA invoice carries the hash of the one before it (PIH), forming a chain
per EGS unit. Because each tablet is its own EGS unit, **each tablet owns its own
chain and its own counter** — chains are never shared, merged, or restarted.

⚠️  The canonicalization and hashing rules below must be validated against
ZATCA's official SDK before production. Getting C14N subtly wrong produces a
hash that looks fine locally and is rejected by Fatoora. See README.
"""

from __future__ import annotations

import base64
import hashlib
from dataclasses import dataclass

# ZATCA's documented seed for the first invoice in a chain: the SHA-256 hex
# digest of the string "0", base64-encoded.
INITIAL_PIH = base64.b64encode(
    hashlib.sha256(b"0").hexdigest().encode("ascii")
).decode("ascii")


def invoice_hash(canonical_xml: bytes) -> str:
    """base64(SHA-256(canonicalised invoice XML))."""
    return base64.b64encode(hashlib.sha256(canonical_xml).digest()).decode("ascii")


@dataclass
class ChainState:
    """Where one device's invoice chain currently stands.

    `icv` is the counter for the **next** invoice. It must increase by exactly
    one per invoice and must never be reused: a gap or a repeat breaks the chain
    and ZATCA rejects everything after it.
    """

    icv: int = 1
    pih: str = INITIAL_PIH

    def advance(self, new_hash: str) -> "ChainState":
        return ChainState(icv=self.icv + 1, pih=new_hash)


class ChainError(RuntimeError):
    pass


def validate_chain(entries: list[tuple[int, str, str]]) -> None:
    """Check a device's issued invoices form an unbroken chain.

    `entries` is (icv, pih, invoice_hash) ordered by icv. Raises on the first
    problem — this is what a reconciliation job runs to prove a device's history
    is intact before anyone tries to explain a rejection to the tax authority.
    """
    if not entries:
        return

    expected_icv = entries[0][0]
    if expected_icv != 1:
        raise ChainError(f"chain starts at ICV {expected_icv}, expected 1")

    prev_hash = INITIAL_PIH
    for icv, pih, own_hash in entries:
        if icv != expected_icv:
            raise ChainError(
                f"ICV {icv} out of sequence, expected {expected_icv} "
                "(a gap or reuse breaks the chain)"
            )
        if pih != prev_hash:
            raise ChainError(
                f"ICV {icv}: PIH does not match the previous invoice's hash"
            )
        prev_hash = own_hash
        expected_icv += 1

File: app/lib/zatca/hashing.dart
/// Invoice hashing and the per-device hash chain.
///
/// Dart port of `zatca/zatca/hashing.py`. Every ZATCA invoice carries the
/// hash of the one before it (PIH), forming a chain per EGS unit. Because
/// each tablet is its own EGS unit, **each tablet owns its own chain and its
/// own counter** — chains are never shared, merged, or restarted. The chain
/// state lives in the `device` row and advances inside the same transaction
/// that closes the sale.
library;

import 'dart:convert';
import 'dart:typed_data';

import 'package:pointycastle/digests/sha256.dart';

Uint8List sha256Bytes(List<int> data) =>
    SHA256Digest().process(Uint8List.fromList(data));

String _hex(Uint8List bytes) =>
    bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();

/// ZATCA's documented seed for the first invoice in a chain: the SHA-256 hex
/// digest of the string "0", base64-encoded.
final String initialPih =
    base64Encode(ascii.encode(_hex(sha256Bytes(ascii.encode('0')))));

/// base64(SHA-256(canonicalised invoice XML)).
String invoiceHash(List<int> canonicalXml) =>
    base64Encode(sha256Bytes(canonicalXml));

/// Where one device's invoice chain currently stands.
///
/// `icv` is the counter for the **next** invoice. It must increase by exactly
/// one per invoice and must never be reused: a gap or a repeat breaks the
/// chain and ZATCA rejects everything after it.
class ChainState {
  const ChainState({this.icv = 1, this.storedPih});

  final int icv;

  /// Null on a device that has issued nothing yet — [pih] then seeds the
  /// chain, rather than a caller having to know the seed value.
  final String? storedPih;

  String get pih => storedPih ?? initialPih;

  ChainState advance(String newHash) =>
      ChainState(icv: icv + 1, storedPih: newHash);
}
